stablehlo/dialect/Base.cpp

```diff
@@ -261,67 +261,96 @@ FailureOr<std::pair<int64_t, int64_t>> inferMergedDimAndBound(
   return std::make_pair(size, bound);
 }
 
-// TODO(zhouxin) Refactor to better handle errors and return single type
-LogicalResult inferMostSpecificType(
+// Inference rules for conditional branches (lhs/rhs are commutative):
+//       Dim of lhs     Dim of rhs      Infer
+//  c0:  X              X               X
+//  c1:  X              ?               ?
+//  c2:  X              ?, B(>=X)       ?, B(>=X)
+//  c3:  X              ?, B(<X)        ?, B(<X)
```

Comment:
From what I can see, [the HLO code mentioned in PR's description](https://github.com/tensorflow/tensorflow/blob/d5b57ca93e506df258271ea00fc29cf98383a374/tensorflow/compiler/xla/service/shape_inference.cc#L2999) only allows `B = X`, but not any other values for B. What are the pros and cons of doing the same thing as HLO vs being more general as proposed in this pull request?

Reply:
I give a detailed proposal in the PR description, which answer this question as well, PTAL.